Make the documentation of completionItem larger

tamuratak opened the pull request:
This is a PR to make the documentation of completionItem larger, which solves #14165 along with commits by @sandy081 solving #67076.

Before:

![2019-02-25 21 32 17](https://user-images.githubusercontent.com/10665499/53337474-dd028500-3944-11e9-84f3-06eaf3340b32.png)


After:

![2019-02-25 21 28 23](https://user-images.githubusercontent.com/10665499/53337425-b7757b80-3944-11e9-8a97-c231a2c8bb9f.png)

At various positions:

![feb-25-2019 21-34-55](https://user-images.githubusercontent.com/10665499/53337631-4a161a80-3945-11e9-8ba6-493196306efb.gif)

msftclas commented:
[![CLA assistant check](https://cla.opensource.microsoft.com/pull/badge/not_signed)](https://cla.opensource.microsoft.com/Microsoft/vscode?pullRequest=69363) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.opensource.microsoft.com/Microsoft/vscode?pullRequest=69363) before we can accept your contribution.<br/><br/>:x: tamuratak [sign now](https://cla.opensource.microsoft.com/Microsoft/vscode?pullRequest=69363)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.opensource.microsoft.com/check/Microsoft/vscode?pullRequest=69363) it.</sub>

tamuratak commented:
I found that this PR sovles #29126 too.

Astrantia commented:
@sandy081 @joaomoreno Why not merge this for the next release?

joaomoreno commented on src/vs/editor/contrib/suggest/suggestWidget.ts:
```diff
@@ -961,12 +959,95 @@ export class SuggestWidget implements IContentWidget, IListVirtualDelegate<Compl
+			return null;
+		}
+		const tmp = Math.max(leftDetailWidth, rightDetailWidth, this.listWidth);
+		return Math.min(tmp, editorWidth * 0.66);
```
This should be rounded, since it will cause fractional numbers.

tamuratak commented:
I would like to resolve the conflict if needed.

joaomoreno commented:
Sure, go ahead :+1:

joaomoreno commented:
Don't forget this:

> Looks good, **but I would put this behind a setting and leave the default as is.**

jrieken commented:
Not a fan of the approach. This will only look good when there is loads of contents and this will look bad when there is little content. Can't we have something that adapts (within min/max bounds) to the size of the rendered contents?

tamuratak commented:
Just like the solution in #69388, allow large details only when code blocks exist.

![Aug-14-2019 08-52-18](https://user-images.githubusercontent.com/10665499/62985280-97ddfb00-be71-11e9-833c-5b27283c8c00.gif)

tamuratak commented:
With `vs/base/common/marked/marked`, the maximum length of lines in code blocks is used to set the width of details if they exist. If not, the width is set `500px` when this feature is enabled. When it is disabled, the width is set `330px`.

Enabled:

![2019-08-18 16 15 35](https://user-images.githubusercontent.com/10665499/63221448-102b2000-c1d4-11e9-8372-c355d446f945.png)

![2019-08-18 16 24 56](https://user-images.githubusercontent.com/10665499/63221507-c4c54180-c1d4-11e9-9aaf-8935f7ec9856.png)

Disabled:

![2019-08-18 16 26 25](https://user-images.githubusercontent.com/10665499/63221517-f76f3a00-c1d4-11e9-8926-52d388ea5bc2.png)

@joaomoreno, fixed reacting to the setting change. Fixed the size of `No suggestions`.

![2019-08-18 16 53 09](https://user-images.githubusercontent.com/10665499/63221765-c1cc5000-c1d8-11e9-8300-29e3cc4d0f7e.png)

MOZGIII commented:
This might not be perfect as it is, however it definitely solves a pain point if the screenshots represent the reality. What if this is merged in as is, and then polished to automatically adapt to the side alter? It looks like users can switch to this new mode only if they want to, so this should be good enough for now.

awulkan commented:
This looks good enough to me for now. Right now the documentation window is barely readable in a lot of cases, and I'd hate to see a potential solution be stuck for years because it's not 100% perfect. It's better than what we currently have, and I believe it can be improved through incremental updates later on.

wompeter commented:
These tiny popup windows have been a major usability issue for literally years, with people resorting to hacking the CSS and re-hacking it every update.  I can't believe the reluctance to deal with the problem when there are solutions like this waiting to go.

jrieken commented:
I think we have been pretty clear that the approach of two static sizes (large and not so large) is not what we want. We are looking for dynamic sizing that tries to fit the contents of the suggestion details widget, actually we try to have the same for hovers as well.

tamuratak commented:
I heard it for the first time.

> I think we have been pretty clear that the approach of two static sizes (large and not so large) is not what we want.

Do you mean that you have started implementing a dynamic sizing suggestion widget? Can I try it on a certain branch?

> We are looking for dynamic sizing that tries to fit the contents of the suggestion details widget, actually we try to have the same for hovers as well.

MOZGIII commented:
> I think we have been pretty clear that the approach of two static sizes (large and not so large) is not what we want. We are looking for dynamic sizing that tries to fit the contents of the suggestion details widget, actually we try to have the same for hovers as well.

Could you elaborate more on the desired approach? Or link us to a more in-depth discussion on the matter?

tamuratak commented:
@jrieken  in the following gif, for example, are you saying that the width of details should shrink to the actual length of inside texts if they are short?

However, if the width of suggestions details changes too dynamically, it will annoy users.

![Apr-15-2020 22-59-36](https://user-images.githubusercontent.com/10665499/79346189-062d0d00-7f6d-11ea-81a6-f76a621c9239.gif)

Moreover, when the width of suggestions changes dynamically, it is difficult to determine the correct position of it. In some cases, we have to change the position dynamically to prevent it going out of the window. And, it will annoy users.

wompeter commented:
> I think we have been pretty clear that the approach of two static sizes (large and not so large) is not what we want. We are looking for dynamic sizing that tries to fit the contents of the suggestion details widget, actually we try to have the same for hovers as well.

Is anyone actually looking though?  Because this issue has been open in various forms since at least 2016.
https://github.com/microsoft/vscode/issues/14165

Debugging in vscode is so painful because of this.
